### ISSUE TITLE
Training NER with gold annotations containing blank spaces and ASCII chars - to clean or not to clean?

### ISSUE THREAD

**Issue by kormilitzin** [explosion/spaCy#4980]
I have merged several gold annotated datasets (from the same corpora) for NER training, however the problem is that some examples contain blank spaces, tabs, new lines (\n, \t) and post-OCR artefacts. For example:

train_data = [
    ("Uber blew through $1 million a week", [(0, 4, 'ORG')]),
    ("\n\nAndroid Pay expands to \nCanada", [(2, 13, 'PRODUCT'), (24, 31, 'GPE')]),
    ("Spotify steps up \n\nAsia expansion", [(0, 8, "ORG"), (19, 23, "LOC")])]

I've always assumed that the tokeniser should not be very much bothered by these, however I observed a slight improvement in F1 metric while 'cleaning' white spaces, newlines and adjusting the annotations character offsets:

train_data = [
    ("Uber blew through $1 million a week", [(0, 4, 'ORG')]),
    ("Android Pay expands to Canada", [(0, 11, 'PRODUCT'), (23, 30, 'GPE')]),
    ("Spotify steps up Asia expansion", [(0, 8, "ORG"), (17, 21, "LOC")])]

** My question 1** The improvement in F1 that I see, could it be an artefact of my training, or it is advised not to mix gold annotations with newlines and other formatting sequences?

** My question 2** Is there a simple way to remove newlines and white spaces from gold annotations? (remove and adjust offset values of characters).

Thanks.

(I'm using Python 3.7 with spaCy 2.2.3 on Windows 10 and anaconda)

**Comment by adrianeboyd** [comment]
Your training data should look like the data you want to process with your model. If you preprocess your training data to clean up whitespace, you should also plan to preprocess the data you're using the future or the performance on the new data will be worse than you expect.

The slight improvement with preprocessed data is probably real, but should be less noticeable as you train on more data. Spacy treats extra whitespace like `\n` as a token in the sentence, so the examples `"\n\nAndroid Pay expands to \nCanada"` and `"Android Pay expands to Canada"` are not quite equivalent while training.

We don't have any built-in utilities to do this kind of preprocessing, sorry.

If you decide not to preprocess and you're still noticing worse performance on the training data with extra whitespace, I'd recommend doing some kind of data augmentation where you generate additional training instances by moving or inserting whitespace from existing training instances.

**Comment by kormilitzin** [comment]
Hi @adrianeboyd 

Thanks a lot for your detailed answer! I'm also using Prodigy to collect gold examples (which are cleaned), so you are right, that the model is still confused by blending clean and not cleaned, but because of the numbers (>1,000 examples) the differences are quite small. That perfectly confirms my instinct to be systematic (re cleaning).

BTW, back in time (several weeks ago), I had a question/problem, that my spacy CLI training was failing and eventually I found that some tokens contained `'\n` character in the way it was not possible to split them (for example 'award\n-winning') and it crashed the training loop with other examples without `\n` chars. It all makes sense now.

Thanks again!
